### ISSUE TITLE
Support for RTL cursive languages? (Arabic in particular)

### ISSUE THREAD

**Issue by PeacefulGrapes** [rhaleblian/dslibris#24]
I've been using this app for reading and it's been great but I was wondering if it can view epub files that are in Arabic language (for a bit of context; the Arabic language is an RTL, cursive language).

I managed to view the file using an Arabic font but the letters were from left to right they weren't conjoined as they should be.

[Here](https://drive.google.com/file/d/18X2xsxs-RJS8eA6cFJJzQKhKnO35Dj5c/view?usp=sharing) is a picture of how it appears in the dslibris app (after using the Arabic font) and [here](https://drive.google.com/file/d/1tR5z21Elx44PAHo26oufOq9YTPe8sRFu/view?usp=sharing) is another one showing how it should be (the app used is Lithium for Android and [here](https://drive.google.com/file/d/1lwEyuh5Syojpq-kBy_l8uqJ65zwjqZU-/view?usp=sharing) is the epub file in Arabic that I'm doing my tests on).

The font file I've used to show the letters in dslibris is [DroidNaskh.ttf](https://drive.google.com/file/d/1Bk3kHL_PTT-IVcloDFwyZ1cnbWnJKH6v/view?usp=sharing) (link provided), without it the letters would show as squares.

If there's any additional information you'd like to know about the case or the language itself that could help you add Arabic support to the app I'd be glad to help.

**Comment by rhaleblian** [comment]
Good suggestion.  Arabic was on my mind years ago when working on the layout algorithm.  Perhaps you could be a tester for RTL.

**Comment by rhaleblian** [comment]
As a side note, another case is traditional Japanese, which is vertical and RTL.

**Comment by PeacefulGrapes** [comment]
> Good suggestion. Arabic was on my mind years ago when working on the layout algorithm. Perhaps you could be a tester for RTL.

I'd be glad to help make it happen, and I sincerely appreciate your work.

> As a side note, another case is traditional Japanese, which is vertical and RTL.

Yeah but I don't think it's cursive so I think Arabic would require some additional work.

**Comment by rhaleblian** [comment]
You're correct.

**Comment by EL-S** [comment]
I'd be happy to test for Farsi

**Comment by PeacefulGrapes** [comment]
> I'd be happy to test for Farsi

Adding support to Arabic automatically extends to include other languages that share the same alphabet and writing characteristics such as Persian, so that's nothing to worry about.

**Comment by rhaleblian** [comment]
The general approach would be to negate the advances provided by the font and allow initialization of each line to be onto the opposite side of the page, all controllable by a "right-to-left" mode alternate to "left-to-right" mode.  Look to see if a font provides any hint about its direction.

**Comment by rhaleblian** [comment]
Using the book and font here, i get mostly black pages with a few printed characters outside the range of Arabic.  Need to turn the debug drawing on to see what's up.
One potential issue is possible variation in "UTF-8 or not", though i can't see why this would ever be user-specific.  Seems like using UTF-16 or UTF-32 for Arabic is not unreasonable because of where Arabic is placed in the UTF codepages.

**Comment by PeacefulGrapes** [comment]
I'm using [this](https://github.com/rhaleblian/dslibris/releases/tag/v1.5.1) version (v1.5.1).

**Comment by rhaleblian** [comment]
I'm on 1.6.0 with some dev changes, which isn't yet released.  That's very possibly the reason.

**Comment by PeacefulGrapes** [comment]
I don't mean to rush things up but is there any news concerning Arabic development?

**Comment by rhaleblian** [comment]
It's not in the cards from my viewpoint.
There aren't any other individuals involved in this project, but this would be a great place for someone to contribute.

**Comment by PeacefulGrapes** [comment]
Yeah you're right especially with the DS having gotten so old. Anyway, thanks for the great app.

**Comment by rhaleblian** [comment]
ref: #31

**Comment by rhaleblian** [comment]
Out-of-scope.  Work has been done seeing if a 3DS app could do this; yes it could.
